File: host/port.py
import time
import serial
import serial.tools.list_ports
import logging


def find(serial_device):
    device_list = serial.tools.list_ports.comports()
    device_string = ''
    for device in device_list:
        if device.product == serial_device and device.manufacturer == 'Plasmatronics':
            if device_string == '':
                device_string = device.device
            else:
                raise Exception('Found Multiple Devices "' + serial_device + '"')
    if device_string == '':
        raise Exception(('Could not Find Device "' + serial_device + '"'))
    return device_string

class port:
    def __init__(self, serial_device):
        self.device = serial_device
        serialPort = find(serial_device)
        logging.debug('Using ' + serialPort + ' for ' + self.device)
        self.ser = serial.Serial(
            port=serialPort,
            baudrate=9600,
            bytesize=serial.EIGHTBITS,
            stopbits=serial.STOPBITS_ONE,
            parity=serial.PARITY_NONE,
            timeout=3)  # default timeout for reading response in seconds 
        self.rx_buffer = bytearray()  # Temporary storage for incoming data
        time.sleep(0.5) # allow time for initialisation of port to complete 
        self.ser.flushOutput()
        self.ser.flushInput()

    def write(self, msg):
        self.ser.flushOutput()
        self.ser.flushInput()
        self.ser.write(msg)

    def read(self):
        # global buffer
        data = self.ser.read(self.ser.in_waiting)  # Read all available bytes
        if data:
            self.rx_buffer.extend(data)  # Append new data to buffer
            while b'\x00' in self.rx_buffer:  # Check if there's a NUL-terminated string
                nul_index = self.rx_buffer.index(b'\x00') + 1
                message = self.rx_buffer[:nul_index]
                self.rx_buffer = self.rx_buffer[nul_index + 1:]  # Remove processed data
                return message  # Return the first complete message
        return None  # No complete message yet

    def close(self):
        self.ser.close()

File: host/javascript/port.js
const { SerialPort } = require('serialport');
const { list } = require('@serialport/list');

async function find(serialDevice) {
    const deviceList = await list();
    let deviceString = '';
    let foundCount = 0;
    
    for (const device of deviceList) {
        if (device.product === serialDevice && device.manufacturer === 'Plasmatronics') {
            if (deviceString === '') {
                deviceString = device.path;
                foundCount++;
            } else {
                throw new Error(`Found Multiple Devices "${serialDevice}"`);
            }
        }
    }
    
    if (deviceString === '') {
        throw new Error(`Could not Find Device "${serialDevice}"`);
    }
    return deviceString;
}

class Port {
    constructor(serialDevice) {
        this.device = serialDevice;
        this.rxBuffer = Buffer.alloc(0); // Temporary storage for incoming data
    }

    async init() {
        const serialPortPath = await find(this.device);
        console.debug(`Using ${serialPortPath} for ${this.device}`);
        
        this.ser = new SerialPort({
            path: serialPortPath,
            baudRate: 9600,
            dataBits: 8,
            stopBits: 1,
            parity: 'none',
            autoOpen: false
        });

        // Set up event handlers
        this.ser.on('open', () => {
            console.debug('Serial port opened');
        });

        this.ser.on('error', (err) => {
            console.error('Serial port error:', err);
        });

        this.ser.on('data', (data) => {
            this.handleData(data);
        });

        // Open the port
        await new Promise((resolve, reject) => {
            this.ser.open((err) => {
                if (err) reject(err);
                else resolve();
            });
        });

        // Allow time for initialization
        await new Promise(resolve => setTimeout(resolve, 500));
        
        await this.flush();
        return this;
    }

    async flush() {
        return new Promise((resolve) => {
            this.ser.flush((err) => {
                if (err) console.error('Flush error:', err);
                resolve();
            });
        });
    }

    async write(msg) {
        await this.flush();
        
        return new Promise((resolve, reject) => {
            this.ser.write(msg, (err) => {
                if (err) reject(err);
                else resolve();
            });
        });
    }

    handleData(data) {
        // Append new data to buffer
        this.rxBuffer = Buffer.concat([this.rxBuffer, data]);
        
        // Check for NUL-terminated strings
        const nulIndex = this.rxBuffer.indexOf(0x00);
        if (nulIndex !== -1) {
            // Extract message including the NUL terminator
            const message = this.rxBuffer.subarray(0, nulIndex + 1);
            // Remove processed data from buffer
            this.rxBuffer = this.rxBuffer.subarray(nulIndex + 1);
            
            // Emit the message (you might want to use events or callbacks)
            if (this.onMessage) {
                this.onMessage(message);
            }
        }
    }

    // Alternative read method that returns a promise
    read(timeout = 3000) {
        return new Promise((resolve, reject) => {
            const timeoutId = setTimeout(() => {
                reject(new Error('Read timeout'));
            }, timeout);

            const messageHandler = (message) => {
                clearTimeout(timeoutId);
                this.removeListener('message', messageHandler);
                resolve(message);
            };

            this.on('message', messageHandler);
        });
    }

    // Event emitter methods (simplified)
    on(event, callback) {
        if (event === 'message') {
            this.onMessage = callback;
        }
    }

    removeListener(event, callback) {
        if (event === 'message' && this.onMessage === callback) {
            this.onMessage = null;
        }
    }

    async close() {
        return new Promise((resolve) => {
            this.ser.close((err) => {
                if (err) console.error('Close error:', err);
                resolve();
            });
        });
    }
}

// Usage example:
async function main() {
    try {
        const port = new Port('YourDeviceName');
        await port.init();
        
        // Set up message handler
        port.on('message', (message) => {
            console.log('Received message:', message);
        });
        
        // Write data
        await port.write(Buffer.from([0x01, 0x02, 0x03]));
        
        // Read with timeout
        try {
            const response = await port.read(5000);
            console.log('Response:', response);
        } catch (err) {
            console.log('No response received within timeout');
        }
        
        await port.close();
    } catch (err) {
        console.error('Error:', err);
    }
}

module.exports = { Port, find };